stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/telecom/TelecomCall.kt

```diff
@@ -0,0 +1,178 @@
+
+    val attributes: CallAttributesCompat
+        get() = CallAttributesCompat(
+            displayName = streamCall.state.createdBy.value?.userNameOrId ?: "Unknown",
+            address = Uri.parse("https://getstream.io/video/join/${streamCall.cid}"),
+            direction = if (state == TelecomCallState.INCOMING) {
+                CallAttributesCompat.DIRECTION_INCOMING
+            } else {
+                CallAttributesCompat.DIRECTION_OUTGOING
+            },
+            callType = mediaType,
+        )
+
+    val mediaType: Int
+        get() = if (streamCall.hasCapability(OwnCapability.SendVideo)) {
+            CallAttributesCompat.CALL_TYPE_VIDEO_CALL
+        } else {
+            CallAttributesCompat.CALL_TYPE_AUDIO_CALL
+        }
+
+    var callControlScope: CallControlScope? = null
+        set(value) {
+            field = value
+            value?.let(::publishDevices)
+        }
+
+    var deviceListener: DeviceListener? = null
+        set(value) {
+            field = value
+            value?.let(::collectDevices)
+        }
+
+    private val localScope = CoroutineScope(parentScope.coroutineContext + Job())
+    private val devices = MutableStateFlow<Pair<List<StreamAudioDevice>, StreamAudioDevice>?>(null)
+
+    fun updateState(newState: TelecomCallState) {
+        val joined = TelecomCallState.IDLE to TelecomCallState.ONGOING
+        val answered = TelecomCallState.INCOMING to TelecomCallState.ONGOING
+        val accepted = TelecomCallState.OUTGOING to TelecomCallState.ONGOING
+        val transition = state to newState
+        state = newState
+
+        StreamLog.d(TELECOM_LOG_TAG) { "[updateState] New state: $state" }
+
+        callControlScope?.let {
+            it.launch {
+                when (transition) {
+                    joined, accepted -> it.setActive().let { result ->
+                        StreamLog.d(TELECOM_LOG_TAG) { "[updateState] Set active: $result" }
+                    }
+
+                    answered -> it.answer(mediaType).let { result ->
+                        StreamLog.d(TELECOM_LOG_TAG) { "[updateState] Answered: $result" }
+                    }
+                }
+            }
+        }
+    }
+
+    suspend fun handleTelecomEvent(event: TelecomEvent) {
+        StreamLog.d(TELECOM_LOG_TAG) { "[TelecomCall#handleTelecomEvent] event: $event" }
+
+        when (event) {
+            TelecomEvent.ANSWER -> {
+                streamCall.accept()
+                streamCall.join()
+            }
+            TelecomEvent.DISCONNECT -> {
+                streamCall.leave()
+            }
+            TelecomEvent.SET_ACTIVE -> {
+                streamCall.join()
+            }
+            TelecomEvent.SET_INACTIVE -> {
+                streamCall.leave()
+            }
+        }
+    }
+
+    fun cleanUp() {
+        StreamLog.d(TELECOM_LOG_TAG) { "[TelecomCall#cleanUp]" }
+
+        runBlocking {
+            disconnect()
+            localScope.cancel()
+        }
```

Comment:
I think we should avoid using `runBlocking` here because we can't guarantee that the `cleanUp` function will be called on an I/O thread from the call site. Additionally, using nested `runBlocking` (the TelecomHandler.cleanUp() calls the function) could lead to an infinite UI freeze if it's invoked on the main thread.